src/components/tl/mlx5/tl_mlx5_team.c

```diff
@@ -169,40 +146,153 @@ static inline ucc_status_t ucc_tl_mlx5_a2a_team_test(ucc_base_team_t *team)
-    ucc_status_t        a2a_status   = UCC_OK;
-    ucc_status_t        mcast_status = UCC_OK;
+    ucc_tl_mlx5_team_t            *tl_team      = ucc_derived_of(team, ucc_tl_mlx5_team_t);
+    ucc_team_t                    *core_team    = UCC_TL_CORE_TEAM(tl_team);
+    ucc_subset_t                   subset       = {.map = UCC_TL_TEAM_MAP(tl_team),
+                                                .myrank = UCC_TL_TEAM_RANK(tl_team)};
+    ucc_status_t                   a2a_status   = UCC_OK;
+    ucc_status_t                   mcast_status = UCC_OK;
+    ucc_tl_mlx5_mcast_coll_comm_t *comm         = NULL;
+    ucc_status_t                   status;
+
+
+    if (tl_team->global_sync_in_progress) {
+        status = ucc_service_coll_test(tl_team->global_sync_req);
+        if (status < 0) {
+            tl_debug(UCC_TL_TEAM_LIB(tl_team),
+                     "failure during service coll exchange: %s",
+                     ucc_status_string(status));
+            return status;
+        }
+        if (UCC_INPROGRESS == status) {
+            return status;
+        }
+
+        ucc_service_coll_finalize(tl_team->global_sync_req);
+
+        tl_team->global_sync_in_progress = 0;
+
+        if (tl_team->mcast_state == TL_MLX5_TEAM_STATE_MCAST_CTX_CHECK &&
+            tl_team->a2a_state == TL_MLX5_TEAM_STATE_ALLTOALL_CTX_CHECK ) {
+            tl_team->a2a_status.global = tl_team->global_status_array[UCC_TL_MLX5_A2A_STATUS_INDEX];
+            tl_team->a2a_state         = TL_MLX5_TEAM_STATE_ALLTOALL_INIT;
+
+            if (tl_team->global_status_array[UCC_TL_MLX5_MCAST_STATUS_INDEX] != UCC_OK) {
+                /* mcast context is not available for some of the team members so we cannot create
+                 * mcast team */
+                tl_debug(UCC_TL_TEAM_LIB(tl_team),
+                         "failure during mcast ctx create, no mcast team support");
+
+                if (tl_team->local_mcast_ctx_ready) {
+                    comm = tl_team->mcast->mcast_comm;
+                    /* release the resources */
+                    if (ibv_dereg_mr(comm->grh_mr)) {
+                        tl_warn(UCC_TL_TEAM_LIB(tl_team),
+                                "ibv_dereg_mr failed");
+                    }
+                    if (ibv_destroy_cq(comm->rcq)) {
+                        tl_warn(UCC_TL_TEAM_LIB(tl_team),
+                                "ibv_destroy_cq failed");
+                    }
+                    ucc_free(comm->params.oob);
+                    ucc_free(comm);
+                    ucc_free(tl_team->mcast);
+                }
+                tl_team->mcast_state = TL_MLX5_TEAM_STATE_MCAST_NOT_AVAILABLE;
+            } else {
+                tl_debug(UCC_TL_TEAM_LIB(tl_team),
+                         "all team members have mcast ctx ready");
+                tl_team->mcast_state = TL_MLX5_TEAM_STATE_MCAST_INIT;
+            }
+
+            return UCC_INPROGRESS;
```

Comment:
why return? don't you need to continue this function?

Reply:
i did not want to add a goto to jump from the UCC_OK on line 238. So we return here and take care of the rest of progress in the next team test call